Reflection_Engine/Query/OutputCount.cs

```diff
@@ -0,0 +1,96 @@
+ * the Free Software Foundation, either version 3.0 of the License, or          
+ * (at your option) any later version.                                          
+ *                                                                              
+ * The BHoM is distributed in the hope that it will be useful,              
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using BH.oM.Reflection;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using BH.oM.Reflection.Interface;
+
+namespace BH.Engine.Reflection
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Interface Methods                         ****/
+        /***************************************************/
+
+        public static int IOutputCount(this IOutput output)
+        {
+            return OutputCount(output as dynamic);
+        }
+
+
+        /***************************************************/
+        /**** public Methods                           ****/
+        /***************************************************/
+
+        public static int OutputCount<T>(this Output<T> output)
+        {
+            return 1;
+        }
+
+        /***************************************************/
+
+        public static int OutputCount<T1, T2>(this Output<T1, T2> output)
+        {
+            return 2;
+        }
+
+        /***************************************************/
+
+        public static object OutputCount<T1, T2, T3>(this Output<T1, T2, T3> output)
+        {
+            return 3;
+        }
+
+        /*************************************/
+
+        public static object OutputCount<T1, T2, T3, T4>(this Output<T1, T2, T3, T4> output)
+        {
+            return 4;
+        }
+
+        /*************************************/
+
+        public static object OutputCount<T1, T2, T3, T4, T5>(this Output<T1, T2, T3, T4, T5> output)
+        {
+            return 5;
+        }
+
+        /***************************************************/
+        /**** Private Methods - fallback                ****/
+        /***************************************************/
+
+        private static object OutputCount(this object output)
+        {
+            Compute.RecordError("Could not find a suitable OutputCount method for the object provided. Method returns -1");
+            return -1;
+        }
```

Comment:
I would recommend a few changes: 
- input can be an IObject instead of an object since this is the only things that the I method accepts. No big deal here but probably good to set good examples everywhere we can.
- this should return an int, instead of an object. Same for some of the methods above.
- you don't have to give up here. There is a simple way to get the result you want through reflection:
```c#
private static object OutputCount(this IOutput output)
{
    if (output == null)
        return 0;
    else
        return output.GetType().GenericTypeArguments.Count();
}
```
This is quite important since we might end up adding a few more `Output` with more generics in the future and forget to update here. 
- Finally, I am not sure avoiding reflection is the right approach here. Why not use a single method instead ?

Reply:
Can get behind using reflection for this. Honestly did not thought of that way of doing it, or I would Have done it. Will replace my code with your method.